plugins/s3_auth/aws_auth_v4.cc

```diff
@@ -0,0 +1,730 @@
+/*
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+/**
+ * @file aws_auth_v4.cc
+ * @brief AWS Auth v4 signing utility.
+ * @see aws_auth_v4.h
+ */
+
+#include <cstring>        /* strlen() */
+#include <ctime>          /* strftime(), time(), gmtime_r() */
+#include <iomanip>        /* std::setw */
+#include <sstream>        /* std::stringstream */
+#include <openssl/sha.h>  /* SHA(), sha256_Update(), SHA256_Final, etc. */
+#include <openssl/hmac.h> /* HMAC() */
+
+#undef AWS_AUTH_V4_DETAILED_DEBUG_OUTPUT
+#ifdef AWS_AUTH_V4_DETAILED_DEBUG_OUTPUT
+#include <iostream>
+#endif
+
+#include "aws_auth_v4.h"
+
+/**
+ * @brief Lower-case Base16 encode a character string (hexadecimal format)
+ *
+ * @see AWS spec: http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+ * Base16 RFC4648: https://tools.ietf.org/html/rfc4648#section-8
+ *
+ * @param in ptr to an input counted string to be base16 encoded.
+ * @param inLen input character string length
+ * @return base16 encoded string.
+ */
+String
+base16Encode(const char *in, size_t inLen)
+{
+  if (nullptr == in || inLen == 0) {
+    return String();
+  }
+
+  std::stringstream result;
+
+  const char *src    = in;
+  const char *srcEnd = in + inLen;
+
+  while (src < srcEnd) {
+    result << std::setfill('0') << std::setw(2) << std::hex << static_cast<int>((*src) & 0xFF);
+    src++;
+  }
+  return result.str();
+}
+
+/**
+ * @brief URI-encode a character string (AWS specific version, see spec)
+ *
+ * @see AWS spec: http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+ *
+ * @param in string to be URI encoded
+ * @param isObjectName if true don't encode '/', keep it as it is.
+ * @return encoded string.
+ */
+String
+uriEncode(const String in, bool isObjectName)
```

Comment:
Sounds good, I like code reuse but thinking more about it I would like to keep it this way and eventually convert it to `TSStringPercentEncode()` when we are ready with the `TSBase<n>Decode()`

Currently `class AwsAuthV4` and all related functions are independent of TS API and considering that `uriEncode()` is simpler and is not much less efficient (pretty much the same except for using `std::string`).